compiler/codegen/library.cpp

```diff
@@ -0,0 +1,196 @@
+  bool startsWithLib = strncmp(executableFilename, "lib", libLength) == 0;
+  if (startsWithLib) {
+    name += &executableFilename[libLength];
+  } else {
+    // libname = executableFilename when executableFilename does not start with
+    // "lib"
+    name = executableFilename;
+  }
+  fileinfo makefile;
+  // TODO: alter location to use generated library directory
+  openLibraryHelperFile(&makefile, "Makefile", name.c_str(), false);
+
+  // Save the CHPL_HOME location so it can be used in the other makefile
+  // variables instead of letting them be cluttered with its value
+  std::string cmd = "echo $CHPL_HOME";
+  std::string chplHome = runCommand(cmd);
+  fprintf(makefile.fptr, "CHPL_HOME = %s\n", chplHome.c_str());
+
+  // TODO: compileline --includes-and-defines adds -I. to the list
+  // automatically.  If the library is in a different directory from the
+  // Makefile that is using this one, that won't be sufficient to find the
+  // header file.  I will handle this with the output directory change
+  std::string cflags = getCompilelineOption("cflags");
+  cflags.erase(cflags.length() - 1); // remove trailing newline
+  std::string includes = getCompilelineOption("includes-and-defines");
+  fprintf(makefile.fptr, "CHPL_CFLAGS = %s %s\n",
+          cflags.c_str(),
+          includes.c_str());
+
+  std::string libraries = getCompilelineOption("libraries");
+  std::string libname = "-l";
+  if (startsWithLib) {
+    // libname = "-l<name>" when executableFilename = "lib<name>"
+    libname += name;
+  } else {
+    // libname = executableFilename plus the extension when executableFilename
+    // does not start with "lib"
+    libname = name;
+    libname += getExtension();
+  }
+  // TODO: adjust for different location for the library, see earlier TODO
+  fprintf(makefile.fptr, "CHPL_LDFLAGS = -L. %s %s \n",
+          libname.c_str(),
+          libraries.c_str());
+
+  std::string compiler = getCompilelineOption("compiler");
+  fprintf(makefile.fptr, "CHPL_COMPILER = %s\n", compiler.c_str());
+
+  std::string linker = getCompilelineOption("linker");
+  fprintf(makefile.fptr, "CHPL_LINKER = %s\n", linker.c_str());
+
+  std::string linkerShared = getCompilelineOption("linkershared");
+  fprintf(makefile.fptr, "CHPL_LINKERSHARED = %s", linkerShared.c_str());
+
+  closeLibraryHelperFile(&makefile);
+}
+
+const char* getExtension() {
+  if (fLibraryCompile) {
+    if (fLinkStyle==LS_DYNAMIC) return ".so";
+    else return ".a";
+  }
+  return "";
+}
+
+void openLibraryHelperFile(fileinfo* fi, const char* name, const char* ext,
+                           bool useTmpDir) {
+  if (ext)
+    fi->filename = astr(name, ".", ext);
+  else
+    fi->filename = astr(name);
+
+  if (useTmpDir) {
+    fi->pathname = genIntermediateFilename(fi->filename);
+  } else {
+    fi->pathname = astr(fi->filename);
+  }
+  openfile(fi, "w");
+}
+
+void closeLibraryHelperFile(fileinfo* fi, bool beautifyIt) {
+  closefile(fi->fptr);
```

Comment:
I'm not so sure it makes sense to beautify a Makefile?

Reply:
Mostly thinking about switching the header to use it as well.  I can send false in for this case.